### ISSUE TITLE
Drop support for ruby 2.3

### ISSUE THREAD

**Issue by deivid-rodriguez** [rubygems/bundler#7586]
I'd like to drop support for ruby 2.3, maybe in the upcoming 2.2 release, maybe right after.

**Comment by colby-swandale** [comment]
I fear dropping Ruby 2.3 support in a feature release would cause a fair amount of confusion as Bundler promises to not break backwards compatibility for non major version releases.

Is there something Ruby 2.3 is (not) doing that's causing a burden or preventing a feature being added?

**Comment by deivid-rodriguez** [comment]
Nope, this is not our policy. Have a look at https://github.com/rubygems/bundler/blob/master/doc/POLICIES.md#compatibility-guidelines. In particular, "Unsupported Ruby versions will be dropped in the first Bundler minor release after support ends.". I think this is a very common practice, and that's why we chose it.

**Comment by deivid-rodriguez** [comment]
It is true that if we do this, we'll cause CIs still running on ruby 2.3 and using unconstrained `gem install bundler` to fail, because of https://github.com/rubygems/rubygems/issues/3089. We should try fixing that issue, but even if we fix it and release it in rubygems, people will still to pick up the fix with `gem update --system`, so... it's hard.

I'm fine with waiting more and trying to improve this area of rubygems first :+1:.

**Comment by colby-swandale** [comment]
I'm super confused by that policy, it's not inline to many past discussions I've had with the core team around how Bundler handles EOL rubies. One of the big points for Bundler v2.0 was so we could drop 1.8 support.

@indirect is https://github.com/rubygems/bundler/blob/master/doc/POLICIES.md#compatibility-guidelines right? It sounds counter to the no breaking changes policy?

**Comment by indirect** [comment]
@colby-swandale I think it lines up with the timeline we discussed, but not the version numbers. 

@deivid-rodriguez this POLICIES document was written before the latest round of version number planning, and I think it is now slightly out of date (sorry! 🙇). my memory of the plan that we had all agreed on was to aim for 1) only dropping old versions in major bundler bumps, and 2) try to release one bundler major version per year, around when old Rubies lose their support, to ensure that Bundler would have a chance to drop unsupported Ruby versions.

It turned out Bundler major version bumps were... uh... more complicated than we had hoped. I am still hoping we can sort out the Bundler versioning issues and execute on the overall plan of one Bundler major version per year, dropping unsupported Rubies at that time. 

In summary:
- The Bundler team has said in public that we will only drop Rubies in major version bumps
- We need some more time to sort out versioning bugs before Bundler 3
- Ruby 2.4 will be unsupported on March 31, 2020
- Let's aim to drop both 2.3 and 2.4 in a Bundler 3.0 release on or after April 1, 2020

How does that sound?

**Comment by deivid-rodriguez** [comment]
As I recall it, we agreed that dropping support for old rubies was not a breaking change, and decided to start dropping support on minors (that's what the current policy says). But since we were changing the previous policy we decided to release bundler 2 dropping support for old rubies as a "one time thing". There's probably a slack record about this, I remember since I was the one who brought up the issue.

Anyways, after releasing bundler 2, we discovered about https://github.com/rubygems/rubygems/issues/3089, so we basically discovered that in our particular case (bundler is installed through rubygems, not through bundler), dropping support for old rubies _is_ a breaking change.

So, I'm good with going back to "only dropping support on majors". Just note that the bundler 2 to bundler 3 transition is something bigger than just cutting master, it's about enabling a whole set of new, breaking behaviour that might not be ready for April.

**Comment by indirect** [comment]
Yeah, the part where Bundler 3 might not be ready by April is the hard part there. :/

Thanks for writing up what you remember! I agree that we originally thought it would be ok to drop rubies in minor releases, and I also agree that today it seems like dropping rubies actually needs to be a major. Glad it turns out we are on the same page after all. :)
